lib/block_view_association.go

```diff
@@ -753,83 +797,132 @@ func (bav *UtxoView) GetUserAssociationsByAttributes(associationQuery *UserAssoc
 	if err != nil {
 		return nil, errors.Wrap(err, "GetUserAssociationsByAttributes: ")
 	}
-	// Store matching associations in a map indexed by AssociationID to
-	// prevent duplicate associations retrieved from the db + UTXO view.
-	associationEntryMap := map[*BlockHash]*UserAssociationEntry{}
-	// First, check the database.
-	dbAssociationEntries, err := bav.GetDbAdapter().GetUserAssociationsByAttributes(associationQuery)
+	// Store matching associations from the db + UTXO view in a slice.
+	var associationEntries []*UserAssociationEntry
+	// First, pull matching association entries from the UTXO view so that
+	// we can track deleted association entries and can properly limit the
+	// number of entries retrieved from the database.
+	newAssociationEntries, deletedAssociationIDs := bav._getUtxoViewUserAssociationEntriesByAttributes(associationQuery)
+	// Store a map of UTXO view association entries to prevent duplicates pulled from the db.
+	utxoViewAssociationEntryMap := make(map[*BlockHash]*UserAssociationEntry)
+	for _, associationEntry := range newAssociationEntries {
+		utxoViewAssociationEntryMap[associationEntry.AssociationID] = associationEntry
+	}
+	// Check the database for matching association entries.
+	dbAssociationEntries, prefixType, err := bav.GetDbAdapter().GetUserAssociationsByAttributes(associationQuery, deletedAssociationIDs)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "GetUserAssociationsByAttributes: ")
 	}
-	for _, dbAssociationEntry := range dbAssociationEntries {
-		associationEntryMap[dbAssociationEntry.AssociationID] = dbAssociationEntry
-	}
-	// Next, check the UTXO view for most recent association entries which
-	// take priority over any with the same AssociationID from the database.
-	newUtxoAssociationEntries, deletedUtxoAssociationEntryIDs := bav._getUtxoUserAssociationEntriesByAttributes(associationQuery)
-	for _, utxoAssociationEntry := range newUtxoAssociationEntries {
-		// Add new association entries found in the UTXO view.
-		associationEntryMap[utxoAssociationEntry.AssociationID] = utxoAssociationEntry
-	}
-	for utxoAssociationEntryID := range deletedUtxoAssociationEntryIDs {
-		// Remove deleted association entries found in the UTXO view.
-		delete(associationEntryMap, utxoAssociationEntryID)
-	}
-	// Convert map to slice.
-	var associationEntries []*UserAssociationEntry
-	for _, matchingAssociationEntry := range associationEntryMap {
-		associationEntries = append(associationEntries, matchingAssociationEntry)
+	// Sort the UTXO view association entries according to the query's specified sort order.
+	sortedUtxoViewAssociationEntries, err := bav.GetDbAdapter().SortUserAssociationEntriesByPrefix(
```

Comment:
I think we want to put all the entries from both the view and the db into one slice and order it so we get the same results after we flush to the db.

Reply:
Spoke offline - this will be addressed in a separate PR.